src/security/KeyData.cc

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (C) 1996-2017 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+#include "anyp/PortCfg.h"
+#include "fatal.h"
+#include "security/KeyData.h"
+#include "SquidConfig.h"
+#include "ssl/bio.h"
+
+/// verify that a private key and cert match
+static bool
+checkPrivateKey(Security::CertPointer &cert, Security::PrivateKeyPointer &pkey)
+{
+#if USE_OPENSSL
+    return X509_check_private_key(cert.get(), pkey.get());
+#elif USE_GNUTLS
+    return true; // TODO find GnuTLS equivalent check
+#else
+    return false;
+#endif
+}
+
+/**
+ * Read certificate from file.
+ * See also: Ssl::ReadX509Certificate function, gadgets.cc file
+ */
+bool
+Security::KeyData::loadX509CertFromFile()
+{
+#if USE_OPENSSL
+    const char *certFilename = certFile.c_str();
+    Ssl::BIO_Pointer bio(BIO_new(BIO_s_file()));
+    if (!bio || !BIO_read_filename(bio.get(), certFilename)) {
+        const auto x = ERR_get_error();
+        debugs(83, DBG_IMPORTANT, "ERROR: unable to load file '" << certFile << "': " << ErrorString(x));
+        return false;
+    }
+
+    X509 *certificate = PEM_read_bio_X509(bio.get(), nullptr, nullptr, nullptr);
+
+#elif USE_GNUTLS
+    const char *certFilename = certFile.c_str();
+    gnutls_datum_t data;
+    Security::ErrorCode x = gnutls_load_file(certFilename, &data);
+    if (x != GNUTLS_E_SUCCESS) {
+        debugs(83, DBG_IMPORTANT, "ERROR: unable to load file '" << certFile << "': " << ErrorString(x));
+        return false;
+    }
+
+    gnutls_pcert_st pcrt;
+    x = gnutls_pcert_import_x509_raw(&pcrt, &data, GNUTLS_X509_FMT_PEM, 0);
+    if (x != GNUTLS_E_SUCCESS) {
+        debugs(83, DBG_IMPORTANT, "ERROR: unable to import certificate from '" << certFile << "': " << ErrorString(x));
+        return false;
+    }
+    gnutls_free(data.data);
+
+    gnutls_x509_crt_t certificate;
+    x = gnutls_pcert_export_x509(&pcrt, &certificate);
+    if (x != GNUTLS_E_SUCCESS) {
+        certificate = nullptr; // paranoid: just in case the *_t ptr is undefined after deinit.
+    }
+#else
+    // to simplify and prevent 'undefined variable errors' in the next code block
```

Comment:
The design for libsecurity code is to have each behavioural action be presented by a single generic function/method with internal ifdef split for how that behaviour is implemented by the supported libraries. Using library-specific methods is what we are trying to move away from.

I would like to move the load-key and load-chain into separate functions. But right now the BIO object used by OpenSSL implementation currently gets in the way of splitting off the chain load as it would have to be passed around between the cert and chain methods when no other library API does so.

If you like I can split off the key loading into a generic method whose OpenSSL implementation is just a call to ReadPrivateKeyFromFile(). I was thinking of leaving that until the ReadPrivateKeyFromFile() code could be cleanly moved into KeyData as the OpenSSL implementation of that method.

Reply:
> The design for libsecurity code is to have each behavioural action be presented by a single generic function/method with internal ifdef split for how that behaviour is implemented by the supported libraries. Using library-specific methods is what we are trying to move away from.

We should not mechanically apply Security API principles to that API _implementation_. Yes, there should be one Security::foo() API for all. However, how that Security::foo() is _implemented_ is a different story. If it makes sense to use library-specific implementations for some good reason, then we should use library-specific implementations.

When there is virtually no shared code or overlap between library-specific code implementing Security::foo(), it is better to `#ifdef` whole method bodies rather than force readers to navigate a complicated maze of `#ifdef` code blocks that depend on each other.

> If you like I can split off the key loading into a generic method whose OpenSSL implementation is just a call to ReadPrivateKeyFromFile().

To address this particular change request, please refactor loadX509CertFromFile() to follow this pattern:

```
Security::KeyData::loadX509CertFromFile()
{
    ... optional shared code ...
#if USE_OPENSSL
    ... OpenSSL-specific code ...
#elif USE_GNUTLS
    ... GnuTLS-specific code ...
#else
    ... code specific to disabled SSL support ...
#endif
    ... optional shared code ...
}
```

The insides of each of those `#ifdef` sections is outside this change request scope.

Reply:
Refactored. 

Notice that layout currently results in public certs being sent twice in TLS handshakes. Because of the required multiple BIO objects. I have left a few XXX in the OpenSSL code for that and a few other checks that seem to be worthwhile, but out of scope for this PR.